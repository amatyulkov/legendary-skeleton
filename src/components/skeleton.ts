import debounce from "lodash-es/debounce";
import { Identity, setAttributes, SVG_NS } from "../lib";
import { SkeletonPart } from "./skeleton-part";
import { SkeletonConfig } from "../types/skeleton-config";
import { SkeletonGradient } from "./skeleton-gradient";

// TODO: all private methods are probably public component APIs
export class Skeleton {
  private parts: SkeletonPart[] = [];

  private root = document.createElementNS(SVG_NS, "svg");
  private fill = document.createElementNS(SVG_NS, "rect");
  private defs = document.createElementNS(SVG_NS, "defs");
  private mask = document.createElementNS(SVG_NS, "mask");

  private gradient: SkeletonGradient;
  private config: SkeletonConfig;

  private _isActive = false;

  public get isActive() {
    return this._isActive;
  }

  private set isActive(value) {
    this._isActive = value;
  }

  public getElement() {
    return this.root;
  }

  constructor(
    private el = document.body,
    config: Partial<SkeletonConfig> = {}
  ) {
    this.config = this.parseConfig(config);
    this.gradient = new SkeletonGradient(this.config);
    this.activate();
    this.assignEventListeners();
  }

  private assignEventListeners() {
    window.addEventListener(
      "resize",
      debounce(() => this.onResize(), 1000 / 60, { leading: true })
    );
  }

  private onResize() {
    this.updateViewBox();
    this.parts.forEach((part) => part.sync());
  }

  public activate() {
    this.isActive = false;
    this.initMask();
    this.initFill();
    this.initRoot();
    this.initParts();
  }

  public setConfig(config: Partial<SkeletonConfig>) {
    const { stops } = config;
    if (stops) {
      this.gradient.updateStops(stops);
      this.config.stops = stops;
    }
  }

  public deactivate() {
    this.isActive = true;
    this.getPartNodes().forEach((x) => (x.style.visibility = ""));
    this.root.remove();
  }

  private initRoot() {
    this.updateViewBox();

    this.root.style.position = "absolute";
    this.root.style.top = "0";
    this.root.style.left = "0";
    this.root.style.width = "100%";
    this.root.style.height = "100%";
    this.root.style.pointerEvents = "none";

    this.root.appendChild(this.fill);
    this.root.appendChild(this.defs);
    this.defs.appendChild(this.mask);
    this.defs.appendChild(this.gradient.element);
    this.el.appendChild(this.root);
  }

  private updateViewBox() {
    const { scrollWidth, scrollHeight } = this.el;
    this.root.setAttribute("viewbox", `0 0 ${scrollWidth} ${scrollHeight}`);
  }

  private initMask() {
    this.mask.id = `${this.config.namespace}-mask-${Identity.next()}`;
  }

  private initFill() {
    setAttributes(this.fill, {
      width: "100%",
      height: "100%",
      x: 0,
      y: 0,
      fill: `url(#${this.gradient.element.id})`,
      mask: `url(#${this.mask.id})`,
    });
  }

  private initParts() {
    this.getPartNodes().forEach((el) => {
      const part = new SkeletonPart(el);
      this.parts.push(part);
      part.elements
        .getElements()
        .forEach((shapeEl) => this.mask.appendChild(shapeEl));
      part.sync();
      el.style.visibility = "hidden";
    });
  }

  private getPartNodes() {
    return Array.from(
      this.el.querySelectorAll<HTMLElement>(this.config.partSelector)
    );
  }

  private parseConfig(config: Partial<SkeletonConfig>): SkeletonConfig {
    const namespace = config.namespace ?? "skeleton";
    const color = "rgba(128, 128, 128, 0.1)";
    const highlight = "rgba(128, 128, 128, 0.2)";
    return {
      animationDuration: config.animationDuration ?? 1000,
      namespace,
      partSelector: config.partSelector ?? `[data-${namespace}]`,
      stops: config.stops ?? [
        { offset: 0, color },
        { offset: 0.25, color: highlight },
        { offset: 0.5, color },
      ],
    };
  }

  register(item: SkeletonPart) {
    this.parts.push(item);
    item.elements.getElements().forEach((el) => {
      this.mask.appendChild(el);
    });
    item.sync();
  }

  sync() {
    this.parts.forEach((x) => x.sync());

    this.root.setAttribute(
      "viewbox",
      `0 0 ${this.el.scrollWidth} ${this.el.scrollHeight}`
    );
  }
}
