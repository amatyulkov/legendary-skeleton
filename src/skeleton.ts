import { Identity, setAttributes, SVG_NS } from "./lib";
import { SkeletonPart } from "./skeleton-part";
import { SkeletonConfig } from "./types/skeleton-config";

export class Skeleton {
  nodes: SkeletonPart[] = [];
  gradient: SVGLinearGradientElement;
  root: SVGElement;
  constructor(el = document.body) {
    this.nodes = [];

    const svg = document.createElementNS(SVG_NS, "svg");
    const mask = document.createElementNS(SVG_NS, "mask");
    const fill = document.createElementNS(SVG_NS, "rect");
    const defs = document.createElementNS(SVG_NS, "defs");
    const maskId = "maskId";
    const { scrollWidth, scrollHeight } = document.body;

    mask.id = maskId;

    const style = window.getComputedStyle(el);

    this.config.highlightWidth = style
      .getPropertyValue("--skeleton--highlight-width")
      .trim();
    this.config.animationDuration = style
      .getPropertyValue("--skeleton--animation-duration")
      .trim();
    this.gradient = this.createGradient();

    setAttributes(fill, {
      width: "100%",
      height: "100%",
      x: 0,
      y: 0,
      fill: `url(#${this.gradient.id})`,
      mask: `url(#${maskId})`,
    });

    svg.classList.add("skeleton__svg");
    svg.setAttribute("viewbox", `0 0 ${scrollWidth} ${scrollHeight}`);
    svg.appendChild(fill);
    svg.appendChild(mask);
    el.appendChild(svg);
    this.root = svg;
    defs.appendChild(this.gradient);
    this.root.appendChild(defs);
  }

  config: Partial<SkeletonConfig> = {};

  // TODO: to a separate class
  createGradient() {
    const gradient = document.createElementNS(SVG_NS, "linearGradient");
    const animateX1 = document.createElementNS(SVG_NS, "animate");
    const animateX2 = document.createElementNS(SVG_NS, "animate");
    const animateY1 = document.createElementNS(SVG_NS, "animate");
    const animateY2 = document.createElementNS(SVG_NS, "animate");

    const stopA = document.createElementNS(SVG_NS, "stop");
    const stopB = document.createElementNS(SVG_NS, "stop");
    const stopC = document.createElementNS(SVG_NS, "stop");

    const animations = [animateX1, animateX2, animateY1, animateY2];
    const stops = [stopA, stopB, stopC];
    const percent = parseFloat(this.config.highlightWidth || "50%"); // TODO: config should not be partial

    setAttributes(animateX1, { attributeName: "x1", values: "-200%; 100%" });
    setAttributes(animateY1, { attributeName: "y1", values: "200%; -100%" });
    setAttributes(animateX2, { attributeName: "x2", values: "-100%; 200%" });
    setAttributes(animateY2, { attributeName: "y2", values: "100%; -200%" });

    setAttributes(stopA, {
      offset: "0%",
      ["stop-color"]: "var(--skeleton--base)",
    });
    setAttributes(stopB, {
      offset: `${percent / 2}%`,
      ["stop-color"]: "var(--skeleton--highlight)",
    });
    setAttributes(stopC, {
      offset: `${percent}%`,
      ["stop-color"]: "var(--skeleton--base)",
    });

    gradient.id = `skeleton-gradient-${Identity.next()}`;

    // Safari needs initial values
    setAttributes(gradient, {
      x1: "-200%",
      x2: "-100%",
      y1: "200%",
      y2: "100%",
    });

    animations.forEach((x) => {
      setAttributes(x, {
        dur: this.config.animationDuration || "1000ms", // TODO: config should not be partial
        repeatCount: "indefinite",
      });
      gradient.appendChild(x);
    });
    stops.forEach((x) => gradient.appendChild(x));

    return gradient as SVGLinearGradientElement;
  }

  register(item: SkeletonPart) {
    this.nodes.push(item);
    item.elements.getElements().forEach((el) => {
      this.root.querySelector("mask")?.appendChild(el); // TODO: store mask definitely as non-optional
    });
    item.sync();
  }

  sync() {
    this.nodes.forEach((x) => x.sync());

    this.root.setAttribute(
      "viewbox",
      `0 0 ${document.body.scrollWidth} ${document.body.scrollHeight}`
    );
  }
}
