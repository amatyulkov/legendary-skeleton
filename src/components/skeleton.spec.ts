import { Skeleton } from "./skeleton";

const getStops = (s: Skeleton) =>
  Array.from(s.getElement().querySelectorAll("stop"));

describe("Skeleton", () => {
  it("Should accept gradient stops", () => {
    expect(() => new Skeleton()).not.toThrow();
  });

  it("Should have default gradient stops", () => {
    const skeleton = new Skeleton();

    expect(getStops(skeleton)).toMatchSnapshot("default stops");
  });

  it("Should allow to overwrite stops", () => {
    const skeleton = new Skeleton();

    expect(getStops(skeleton)).toMatchSnapshot("default stops");

    skeleton.setConfig({
      stops: [
        { offset: 0, color: "red" },
        { offset: 0.1, color: "green" },
        { offset: 0.2, color: "blue" },
        { offset: 0.3, color: "white" },
      ],
    });

    expect(getStops(skeleton)).toMatchSnapshot("updated stops");
  });
});
