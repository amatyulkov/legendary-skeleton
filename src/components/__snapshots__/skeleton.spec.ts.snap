// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton Should allow to overwrite stops: default stops 1`] = `
Array [
  <stop
    offset="0%"
    stop-color="rgba(128, 128, 128, 0.1)"
  />,
  <stop
    offset="25%"
    stop-color="rgba(128, 128, 128, 0.2)"
  />,
  <stop
    offset="50%"
    stop-color="rgba(128, 128, 128, 0.1)"
  />,
]
`;

exports[`Skeleton Should allow to overwrite stops: updated stops 1`] = `
Array [
  <stop
    offset="0%"
    stop-color="red"
  />,
  <stop
    offset="10%"
    stop-color="green"
  />,
  <stop
    offset="20%"
    stop-color="blue"
  />,
  <stop
    offset="30%"
    stop-color="white"
  />,
]
`;

exports[`Skeleton Should have default gradient stops: default stops 1`] = `
Array [
  <stop
    offset="0%"
    stop-color="rgba(128, 128, 128, 0.1)"
  />,
  <stop
    offset="25%"
    stop-color="rgba(128, 128, 128, 0.2)"
  />,
  <stop
    offset="50%"
    stop-color="rgba(128, 128, 128, 0.1)"
  />,
]
`;
